### PR TITLE
[Artifact Logging PR1] Logger Refactoring

### LINKED ISSUE #541: Sensible naming scheme for logging classes
Seeing as:
- The `LoggerCallback`-based classes (e.g. `FileLogger`) have "Logger", not "Callback" in their name.
- Referring to the `LoggerCallback`s as `LoggerCallbacks` feels ungainly and seems confusing, because it's focusing on _how_ they work instead of _what_ they're used for.
- Most documentation and instructional material refers to `LoggerCallback`s as "Loggers", which overloads the term.
- `Trainer` has a `loggers` hparam that takes a list of `LoggerCallback`s, not `Loggers`s.
- We renamed `LoggerBackend` to `LoggerCallback`, but the `Logger` class still uses `backend/backends` as a parameter and attribute name.

I propose that we come up with a more sensible naming scheme. I'm not super opinionated, but I have a few ideas. Perhaps we change `Logger` to `LogRouter`, `LoggingController`, or `LoggingFrontend`. Thoughts @hanlint, @ravi-mosaicml, @siriuslee

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
Logger refactoring, in preparation for [Artifact Logging](https://www.notion.so/Artifact-Logging-API-d3754be3ec874233910cb0c4f23c9d32)

1. Rename `LoggerCallback` to `LoggerDestination`, and update the hparams and trainer init
1. Renamed `log_metric` to `log_data`. Metric is a confusing name, as it can easily be confused with TorchMetrics, and the logger is not limited to TorchMetrics (nor does it support torchmetrics!). Similarly, renamed `metric_fit`, `metric_epoch`, and `metric_batch` to `data_fit`, `data_epoch`, and `data_batch`.
1. Made the logger a required argument of the engine (nobody is instantiating the engine themselves anyways; this was a relic from BYOT).

Closes #541.

### CODE REVIEW

**Review by jbloxham — APPROVED** [review]
This feels like a step in the right direction, but I'm not convinced by all of these renames. How would you feel about:
`LoggerDestination` => `Logger`
`Logger` => `MainLogger`

I think in my ideal world, `MainLogger` would have the same interface as any other `Logger`, so that `MainLogger` wouldn't need to show up in most type annotations in the first place. I like renaming `LoggerDestination` to `Logger` because the latter is much more succinct, and `LoggerDestination` doesn't mean much to me.

File: composer/core/logging/logger.py [hunk 1]
```diff
@@ -49,43 +49,43 @@ class LogLevel(IntEnum):
...
-    def _get_destinations_for_log_level(self, log_level: LogLevel) -> Generator[LoggerCallback, None, None]:
-        for destination in self.backends:
+    def _get_destinations_for_log_level(self, log_level: LogLevel) -> Generator[LoggerDestination, None, None]:
+        for destination in self.destinations:
             if destination.will_log(self._state, log_level):
                 yield destination
 
-    def metric(self, log_level: Union[str, LogLevel], data: Union[TLogData, Callable[[], TLogData]]) -> None:
-        """Log a metric to the :attr:`backends`.
+    def data(self, log_level: Union[str, LogLevel], data: Union[LoggerDataDict, Callable[[], LoggerDataDict]]) -> None:
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/core/logging/logger.py:79]: I'd normally agree, but the built-in python logger class does `log.info`, `log.error`, so we followed the convention here.

File: composer/core/logging/logger.py [hunk 2]
```diff
@@ -49,43 +49,43 @@ class LogLevel(IntEnum):
 
 
 class Logger:
-    """Logger routes metrics to the :class:`.LoggerCallback`. Logger is what users call from within
+    """Logger routes metrics to the :class:`.LoggerDestination`. Logger is what users call from within
     algorithms/callbacks. A logger routes the calls/data to any different number of destination
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/core/logging/logger.py:53]: I ended up rewriting the docstring.

File: composer/core/logging/logger.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/core/logging/logger.py:86]: Ok, fair enough then - my feelings aren't strong here.

File: composer/core/logging/logger.py [hunk 3]
```diff
@@ -98,22 +98,22 @@ def metric(self, log_level: Union[str, LogLevel], data: Union[TLogData, Callable
...
+    def data_epoch(self, data: LoggerDataDict) -> None:
+        """Helper function for ``self.data(LogLevel.EPOCH, data)``"""
+        self.data(LogLevel.EPOCH, data)
 
-    def metric_batch(self, data: Union[TLogData, Callable[[], TLogData]]) -> None:
-        """Helper function for ``self.metric(LogLevel.BATCH, data)``"""
-        self.metric(LogLevel.BATCH, data)
+    def data_batch(self, data: LoggerDataDict) -> None:
+        """Helper function for ``self.data(LogLevel.BATCH, data)``"""
+        self.data(LogLevel.BATCH, data)
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/core/logging/logger.py:120]: Leaving as is per https://github.com/mosaicml/composer/pull/698#discussion_r826416066.

**Pull request merged**